File: src/ec_gen/set_bipart.py
"""
Set Bipartition

A set bipartition of the set [n] = {1,2,3,...,n} is a partition into exactly
two non-empty disjoint subsets B0 and B1 whose union is [n]. Each element is
assigned to either block 0 or block 1.

This module generates all 2^{n-1} - 1 non-trivial bipartitions of an n-element
set using a Gray code order where successive bipartitions differ by moving a
single element from one block to the other.

Reference:
Frank Ruskey. Simple combinatorial Gray codes constructed by
reversing sublists. Lecture Notes in Computer Science, #762,
201-208. Also downloadable from
http://webhome.cs.uvic.ca/~ruskey/Publications/SimpleGray/SimpleGray.html
"""

from typing import Generator


def stirling2nd2(num: int) -> int:
    """
    The `stirling2nd2` function calculates the Stirling number of the second kind for a given integer
    `num` (k = 2) using a recursive approach.

    :param num: The parameter `num` represents the number of elements in a set
    :type num: int
    :return: the Stirling number of the second kind for the given input num.

    Examples:
        >>> stirling2nd2(5)
        15
    """
    if num < 3:
        return 1
    return 1 + 2 * stirling2nd2(num - 1)


def set_bipart(num: int) -> Generator[int, None, None]:
    """
    The function `set_bipart` generates a sequence of moves that partitions a set of size `num` into two
    subsets.

    :param num: The parameter `num` represents the number of elements in the bi-partition
    :type num: int

    Examples:
        >>> num = 5
        >>> blocks = [0] * num + [1]
        >>> print(blocks[1:])
        [0, 0, 0, 0, 1]
        >>> for idx in set_bipart(num):
        ...     old_val = blocks[idx]
        ...     blocks[idx] = 1 - blocks[idx]
        ...     print(blocks[1:], ": Move {} from B{} to B{}".format(idx, old_val, blocks[idx]))
        ...
        [0, 0, 0, 1, 1] : Move 4 from B0 to B1
        [0, 1, 0, 1, 1] : Move 2 from B0 to B1
        [0, 1, 1, 1, 1] : Move 3 from B0 to B1
        [0, 0, 1, 1, 1] : Move 2 from B1 to B0
        [0, 0, 1, 0, 1] : Move 4 from B1 to B0
        [0, 1, 1, 0, 1] : Move 2 from B0 to B1
        [0, 1, 0, 0, 1] : Move 3 from B1 to B0
        [0, 1, 0, 0, 0] : Move 5 from B1 to B0
        [0, 1, 1, 0, 0] : Move 3 from B0 to B1
        [0, 0, 1, 0, 0] : Move 2 from B1 to B0
        [0, 0, 1, 1, 0] : Move 4 from B0 to B1
        [0, 1, 1, 1, 0] : Move 2 from B0 to B1
        [0, 1, 0, 1, 0] : Move 3 from B1 to B0
        [0, 0, 0, 1, 0] : Move 2 from B1 to B0
    """
    yield from gen0(num)


# The lists S(num) and S'(num) satisfy the following properties:
# 1. Successive bipartitions differ in exactly one element.
# 2. first(S(num)) = 0^{num-1}1
# 3. last(S(num)) = 0^{num-1}1
# 4. first(S'(num)) = last(S(num))


def gen0(num: int) -> Generator[int, None, None]:
    """S(num) — forward generator for set bipartitions.

    :param num: The total number of elements in the set
    :type num: int
    :return: a generator yielding element indices to flip between blocks.
    """
    if num < 3:
        return
    yield num - 1
    yield from gen1(num - 1)
    yield num
    yield from neg1(num - 1)


def gen1(num: int) -> Generator[int, None, None]:
    """S'(num) — alternate forward generator for set bipartitions.

    :param num: The total number of elements in the set
    :type num: int
    :return: a generator yielding element indices to flip between blocks.
    """
    if num < 3:
        return
    yield 2
    yield from neg1(num - 1)
    yield num
    yield from gen1(num - 1)


def neg1(num: int) -> Generator[int, None, None]:
    """S'(num) — reverse generator for set bipartitions.

    :param num: The total number of elements in the set
    :type num: int
    :return: a generator yielding element indices to flip between blocks.
    """
    if num < 3:
        return
    yield from neg1(num - 1)
    yield num
    yield from gen1(num - 1)
    yield 2


if __name__ == "__main__":
    import doctest

    doctest.testmod()

File: rust_ai/src/set_bipart.rs
//! Set Bipartition Generator
//!
//! This module generates sequences of moves that partition a set of size `n` into two subsets.
//! A set bipartition divides a set into two disjoint subsets whose union is the original set.
//!
//! The algorithm generates a Gray code for set bipartitions, where each successive
//! bipartition differs by moving exactly one element from one block to the other.

use genawaiter::{sync::gen, yield_};
use lazy_static::lazy_static;
use std::collections::HashMap;
use std::sync::Mutex;

lazy_static! {
    static ref STIRLING_CACHE: Mutex<HashMap<i32, i32>> = Mutex::new(HashMap::new());
}

/// Calculate the Stirling number of the second kind for a given integer `n` (k = 2).
///
/// # Arguments
///
/// * `n` - Number of elements in a set
///
/// # Returns
///
/// The Stirling number of the second kind S(n, 2).
///
/// # Examples
///
/// ```
/// use rust_ai::set_bipart::stirling2nd2;
///
/// assert_eq!(stirling2nd2(5), 15);
/// ```
pub fn stirling2nd2(n: i32) -> i32 {
    if n < 3 {
        1
    } else {
        1 + 2 * stirling2nd2(n - 1)
    }
}

/// Generate a sequence of moves that partitions a set of size `n` into two subsets.
///
/// # Arguments
///
/// * `n` - Number of elements in the bi-partition
///
/// # Returns
///
/// A generator that yields integers representing which element to move.
///
/// # Examples
///
/// ```
/// use rust_ai::set_bipart::set_bipart;
///
/// let n = 5;
/// let mut b = vec![0; (n + 1) as usize];
/// b[n as usize] = 1;
/// println!("{:?}", &b[1..]);
///
/// for x in set_bipart(n) {
///     let old = b[x as usize];
///     b[x as usize] = 1 - b[x as usize];
///     println!("{:?} : Move {} from B{} to B{}", &b[1..], x, old, b[x as usize]);
/// }
/// ```
pub fn set_bipart(n: i32) -> impl Iterator<Item = i32> {
    gen!({
        for x in gen0(n) {
            yield_!(x);
        }
    })
    .into_iter()
}

fn gen0(n: i32) -> impl Iterator<Item = i32> {
    gen!({
        if n < 3 {
            return;
        }
        yield_!(n - 1);
        for x in gen1(n - 1) {
            yield_!(x);
        }
        yield_!(n);
        for x in neg1(n - 1) {
            yield_!(x);
        }
    })
    .into_iter()
}

fn gen1(n: i32) -> impl Iterator<Item = i32> {
    gen!({
        if n < 3 {
            return;
        }
        yield_!(2);
        for x in neg1(n - 1) {
            yield_!(x);
        }
        yield_!(n);
        for x in gen1(n - 1) {
            yield_!(x);
        }
    })
    .into_iter()
}

fn neg1(n: i32) -> impl Iterator<Item = i32> {
    gen!({
        if n < 3 {
            return;
        }
        for x in neg1(n - 1) {
            yield_!(x);
        }
        yield_!(n);
        for x in gen1(n - 1) {
            yield_!(x);
        }
        yield_!(2);
    })
    .into_iter()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_stirling2nd2() {
        assert_eq!(stirling2nd2(1), 1);
        assert_eq!(stirling2nd2(2), 1);
        assert_eq!(stirling2nd2(3), 3); // 1 + 2*1
        assert_eq!(stirling2nd2(4), 7); // 1 + 2*3
        assert_eq!(stirling2nd2(5), 15); // 1 + 2*7
    }

    #[test]
    fn test_set_bipart() {
        let n = 5;
        let moves: Vec<_> = set_bipart(n).collect();

        // Should generate 2^n - 2 moves (all bipartitions except empty sets)
        // let expected_len = 2_i32.pow(n as u32) - 2;
        assert_eq!(moves.len() as i32, 14);

        // Test that moves generate valid bipartitions
        let mut b = vec![0; (n + 1) as usize];
        b[n as usize] = 1;
        let mut bipartitions = vec![b[1..].to_vec()];

        for &x in &moves {
            b[x as usize] = 1 - b[x as usize];
            bipartitions.push(b[1..].to_vec());
        }

        // All bipartitions should be unique
        let mut sorted = bipartitions.clone();
        sorted.sort();
        sorted.dedup();
        assert_eq!(sorted.len(), bipartitions.len());
    }

    #[test]
    fn test_set_bipart_small() {
        let moves: Vec<_> = set_bipart(2).collect();
        assert_eq!(moves.len(), 0); // 2^2 - 2 = 2, but n < 3 returns empty

        let moves: Vec<_> = set_bipart(3).collect();
        assert_eq!(moves.len(), 2);
    }
}
